libweasyl/libweasyl/models/tables.py

```diff
@@ -688,13 +691,27 @@ def default_fkey(*args, **kwargs):
     Column('userid', Integer()),
     Column('csrf_token', String(length=64)),
     Column('additional_data', JSONValuesColumn(), nullable=False, server_default=text(u"''::hstore")),
+    Column('ip_address', String(length=39), nullable=True),
+    Column('user_agent_id', Integer(), nullable=True),
     default_fkey(['userid'], ['login.userid'], name='sessions_userid_fkey'),
+    default_fkey(['user_agent_id'], ['user_agents.user_agent_id'], name='sessions_user_agent_id_fkey'),
 )
 
 Index('ind_sessions_created_at', sessions.c.created_at)
 Index('ind_sessions_userid', sessions.c.userid)
 
 
+user_agents = Table(
+    'user_agents', metadata,
+    Column('user_agent_id', Integer(), primary_key=True, nullable=False),
+    # SHA-256 should be unique enough... I don't think we need to worry (yet) about collisions)
+    Column('user_agent_sha256', String(length=64), unique=True, nullable=False),
```

Comment:
What does this column do?

Reply:
I was intending that to act as the ultimate unique identifier for a user agent. Since I sliced ``request.user_agent[0:1024]``, there's a (potential) risk for a similar user-agent string to be stored. So I SHA-256'd the entire string, in the event an odd-ball UA appears, so we can 'capture' the entire string via the hash.

At least, that was the initial theory behind that column. Not sure how many legitimate UAs are above 1024 characters in length, but I was trying to cover the edge case.

Reply:
On reflection, I might just drop the column entirely, and we can (in a future PR) integrate Google Analytics to do the capturing of UA information if there are any horribly crazy UA strings.

That being said, if we drop the column, would the following be a good way of pulling the ID out of the ``user_agents`` table without indexing the entire UA string? (I'm not in front of a PSQL instance right now, so I can't directly test it).

```
def get_user_agent_id(ua_string=None):
    """
    Store 
    """
    if not ua_string:
        return None
    else:
        # Store/Retrieve the UA String. Max UA length stored is 1024 characters.
        query = d.engine.scalar("""
            INSERT INTO user_agents (user_agent)
            VALUES (%(user_agent)s)
            ON CONFLICT DO
                UPDATE SET user_agent = %(user_agent)s
                WHERE user_agent = %(user_agent)s
            RETURNING user_agent_id
        """, user_agent=ua_string[0:1024])
        return query
```
Explaining the above:
- We don't know (initially) if the UA is present, so try to insert the portion of the UA which is the most relevant (AKA, the first 1024 characters should contain the most relevant portions).
- In the event that UA is already stored, update the UA to be the currently processed UA. This is because:
- We want to return the user_agent_id always so we can bind it to the ``session`` for the user. However, based on the way I am reading the [documentation](https://www.postgresql.org/docs/9.6/static/sql-update.html), the ``RETURNING`` clause doesn't fire unless a row is actually updated. So we would then have to force the row to be updated (even though nothing is really being changed from the standpoint of stored data) to satisfy PSQL's requirements for that clause.

If, for some reason, we needed more granular resolution for the full user agent string, we could revisit the hash method. But after reflection, the above code snippet is what I am planning on changing the existing PR's ``login.process_useragent()`` to, unless we should for some reason store the SHA hash.